pairing/bls12381/gnark/g1.go

```diff
@@ -0,0 +1,143 @@
+//nolint:dupl // unavoidable duplication between g1 and g2
+package gnark
+
+import (
+	"crypto/cipher"
+	"fmt"
+	"io"
+	"math/big"
+
+	bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381"
+	"go.dedis.ch/kyber/v4"
+)
+
+var _ kyber.SubGroupElement = &G1Elt{}
+
+// G1Elt is a wrapper around a G1 point on the BLS12-381 Gnark curve.
+type G1Elt struct{ inner bls12381.G1Jac }
+
+// MarshalBinary returns a compressed point, without any domain separation tag information
+func (p *G1Elt) MarshalBinary() (data []byte, err error) {
+	var g1aff bls12381.G1Affine
+	g1aff.FromJacobian(&p.inner)
+	res := g1aff.Bytes()
+	return res[:], nil
+}
+
+// UnmarshalBinary populates the point from a compressed point representation.
+func (p *G1Elt) UnmarshalBinary(data []byte) error {
+	var g1aff bls12381.G1Affine
+	_, err := g1aff.SetBytes(data)
+	if err != nil {
+		return fmt.Errorf("setting affine representation: %w", err)
+	}
+
+	p.inner.FromAffine(&g1aff)
+	return nil
+}
+
+func (p *G1Elt) String() string { return p.inner.String() }
+
+func (p *G1Elt) MarshalSize() int { return bls12381.SizeOfG1AffineCompressed }
+
+// MarshalTo writes a compressed point to the Writer, without any domain separation tag information
+func (p *G1Elt) MarshalTo(w io.Writer) (int, error) {
+	buf, err := p.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	return w.Write(buf)
+}
+
+// UnmarshalFrom populates the point from a compressed point representation read from the Reader.
+func (p *G1Elt) UnmarshalFrom(r io.Reader) (int, error) {
+	buf := make([]byte, p.MarshalSize())
+	n, err := io.ReadFull(r, buf)
+	if err != nil {
+		return n, err
+	}
+	return n, p.UnmarshalBinary(buf)
+}
+
+func (p *G1Elt) Equal(p2 kyber.Point) bool { x := p2.(*G1Elt); return p.inner.Equal(&x.inner) }
```

Comment:
return false if `p2` is not a `G1Elt`?

Reply:
These cases should never be hit, and if they are, then it might be safer to panic than to accept.